text/2018-04-08-global-and-local-pinning.md

```diff
@@ -0,0 +1,247 @@
+
+### Remove `is_pinned()`
+
+Another small tweak we'll do is remove `epoch::is_pinned()` and `Handle::is_pinned()`. These functions are only used by `crossbeam-deque` to make sure a fence is executed even if the thread is already pinned.
+
+Instead, we'll add functions `epoch::pin_fence()`, `Handle::pin_fence()`, and `Collector::pin_fence()`, which are equivalent to `pin()`, except they guarantee to execute a `SeqCst` fence.
+
+# Drawbacks
+
+* `epoch::default_handle()` is gone. But it doesn't seem to be useful and we have `epoch::pin()` and `epoch::flush()` anyway. We can reintroduce it in the future if a need actually comes up.
+
+* `Handle::clone()` is gone. However, nobody is using it and when it is used, it's used mistakenly. In short, it comes with practically zero benefits and a serious drawback.
+
+* `Handle::pin()` is now unsafe. See the *Alternatives* section for what we can do about it.
+
+# Alternatives
+
+### Add safe `Handle::pin_with()` and remove `Guard::clone()`
+
+We can add a safe alternative to `Guard::pin()`:
+
+```rust
+impl Handle {
+    fn pin_with<F, R>(&self, f: F) -> R
+    where
+        F: FnOnce(&mut Guard) -> R;
+}
+```
+
+This function is safe because it borrows the handle for the duration of pinning, which means it cannot be sent to another thread in the meantime. However, the user mustn't be able to clone the `Guard` and send the handle to another thread afterwards, as in this example:
+
+```rust
+let guard = handle.pin_with(|guard| guard.clone());
+thread::spawn(move || {
+    handle.pin_with(|_| ());
+});
+drop(guard); // Oops, the handle now belongs to another thread!
+```
+
+Therefore, we'd have to remove the `Clone` impl for `Guard`.
+
+### Make `Handle::pin()` safe, but add a lifetime to `Guard`
+
+Another alternative is to add a lifetime to `Guard` so that it becomes `Guard<'a>`. The lifetime `'a` borrows the `Handle` and therefore prevents it from being moved at all, which means it cannot be sent to another thread while such a guard exists.
+
+This way we don't have to remove `Handle::clone()`.
+
+Note that `Collector::pin()` and `epoch::pin()` simply return a `Guard<'static>`:
+
+```rust
+impl Collector {
+    // We're not using a `Handle` - there's nothing to borrow, so we can use `'static`.
+    fn pin(&self) -> Guard<'static>;
+}
+
+impl Handle {
+    // The returned guard borrows `self`, preventing it from being sent to another thread.
+    fn pin<'a>(&'a self) -> Guard<'a>;
+}
+
+// This function uses the default thread-local handle, but it cannot be sent to another thread,
+// so we don't even have to borrow it.
+fn pin() -> Handle<'static>;
+```
+
+The main drawback of this solution is that the lifetime might be at times annoying, but note that the lifetime can be simply elided in most real-world situations.
+
+### Two kinds of handles: `SharedHandle` and `LocalHandle`
```

Comment:
Are there any drawbacks to this alternative besides it being somewhat more complex to implement and use?

Reply:
Not really, that's the only drawback.